Corona-Warn-App/src/main/java/de/rki/coronawarnapp/datadonation/analytics/Analytics.kt

```diff
@@ -1,34 +1,156 @@
+        val request: DonorModule.Request = object : DonorModule.Request {}
+
+        val contributions = donorModules.map {
+            Timber.d("Beginning donation for module: %s", it::class.simpleName)
+            it.beginDonation(request)
+        }
+
+        contributions.forEach {
+            Timber.d("Injecting contribution: %s", it::class.simpleName)
+            it.injectData(ppaDataBuilder)
+        }
+
+        return contributions
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
     suspend fun submitAnalyticsData() {
-//        val request: DonorModule.Request = ...
-//        // Collect data from all donor modules
-//        val contributions = donorModules.map { it.beginDonation(request) }
-//        val ppaContainer: Any = ...
-//
-//        contributions.forEach {
-//            it.injectData(ppaContainer)
-//        }
-//
-//        val success = trySubmission(ppaContainer)
-//        contributions.forEach {
-//            it.finishDonation(success)
-//        }
+        Timber.d("Starting analytics submission")
+
+        val ppaDataBuilder = PpaData.PPADataAndroid.newBuilder()
+
+        val contributions = collectContributions(ppaDataBuilder = ppaDataBuilder)
+
+        val analyticsProto = ppaDataBuilder.build()
+
+        val success = trySubmission(analyticsProto)
+
+        contributions.forEach {
+            Timber.d("Finishing contribution: %s", it::class.simpleName)
+            it.finishDonation(success)
+        }
+
+        if (success) {
+            settings.lastSubmittedTimestamp.update {
+                Instant.now()
+            }
+
+            logger.storeAnalyticsData(analyticsProto)
+        }
+
+        Timber.d("Finished analytics submission success=%s", success)
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToNoUserConsent(): Boolean {
+        return !settings.analyticsEnabled.value
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    suspend fun stopDueToProbabilityToSubmit(): Boolean {
+        val submitRoll = Random.nextDouble(0.0, 1.0)
+        return submitRoll > appConfigProvider.getAppConfig().analytics.probabilityToSubmit
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToLastSubmittedTimestamp(): Boolean {
+        val lastSubmit = settings.lastSubmittedTimestamp.value ?: return false
+        return lastSubmit.plus(Hours.hours(LAST_SUBMISSION_MIN_AGE_HOURS).toStandardDuration()).isAfterNow
+    }
+
+    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
+    fun stopDueToTimeSinceOnboarding(): Boolean {
+        val onboarding = LocalData.onboardingCompletedTimestamp()?.let { Instant.ofEpochMilli(it) } ?: return true
+        return onboarding.plus(Hours.hours(ONBOARDING_DELAY_HOURS).toStandardDuration()).isAfterNow
+    }
+
+    suspend fun submitIfWanted() {
```

Comment:
I guess this is the entry point to the whole thing, right?
Can we sleep better if we guard this with a `Mutex` or are we absolutely sure there will be no overlaps? If yes why are we so sure?

Reply:
i am relatively sure because of the delay implied in the scheduled background work, but i will add a mutex to be extra safe :)